src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -0,0 +1,143 @@
+
+        /// <summary>
+        /// Creates a targeting contextual feature filter.
+        /// </summary>
+        /// <param name="options">Options controlling the behavior of the targeting evaluation performed by the filter.</param>
+        public ContextualTargetingFeatureVariantAssigner(IOptions<TargetingEvaluationOptions> options)
+        {
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        /// <summary>
+        /// Assigns one of the variants configured for a feature based off the provided targeting context.
+        /// </summary>
+        /// <param name="variantAssignmentContext">Contextual information available for use during the assignment process.</param>
+        /// <param name="targetingContext">The targeting context used to determine which variant should be assigned.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns></returns>
+        public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentContext variantAssignmentContext, ITargetingContext targetingContext, CancellationToken cancellationToken)
+        {
+            if (variantAssignmentContext == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext));
+            }
+
+            if (targetingContext == null)
+            {
+                throw new ArgumentNullException(nameof(targetingContext));
+            }
+
+            FeatureDefinition featureDefinition = variantAssignmentContext.FeatureDefinition;
+
+            if (featureDefinition == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext.FeatureDefinition));
+            }
+
+            if (featureDefinition.Variants == null)
+            {
+                throw new ArgumentNullException(nameof(featureDefinition.Variants));
+            }
+
+            FeatureVariant variant = null;
+
+            FeatureVariant defaultVariant = null;
+
+            double cumulativePercentage = 0;
+
+            var cumulativeGroups = new Dictionary<string, double>();
+
+            foreach (FeatureVariant v in featureDefinition.Variants)
+            {
+                if (defaultVariant == null && v.Default)
+                {
+                    defaultVariant = v;
+                }
+
+                TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+
+                if (targetingSettings == null)
+                {
+                    if (v.Default)
+                    {
+                        //
+                        // Valid to omit audience for default variant
+                        continue;
+                    }
+                    else
+                    {
+                        targetingSettings = new TargetingFilterSettings();
```

Comment:
It was to get the validation from TryValidateSettings. It was a bit confusing. I will update `TryValidateSettings` to be able to take null `TargetingFilterSettings` for validation as well to avoid this confusion.